keyboards/melody96/keymaps/zunger/keymap.c

```diff
@@ -0,0 +1,250 @@
+  // get_mods or the like, because this function is called *before* that's updated!
+  static bool shift_held = false;
+  static bool alt_held = false;
+  static bool greek_held = false;
+  static bool cadet_held = false;
+
+  // These are where we remember the values of lock states.
+  static bool shift_lock = false;
+  static uint32_t layer_lock = _QWERTY;
+
+  // Process any modifier key presses.
+  if (keycode == KC_LSHIFT || keycode == KC_RSHIFT) {
+    shift_held = record->event.pressed;
+  } else if (keycode == KC_LALT || keycode == KC_RALT) {
+    alt_held = record->event.pressed;
+  } else if (keycode == KC_GREEK) {
+    greek_held = record->event.pressed;
+  } else if (keycode == KC_CADET) {
+    cadet_held = record->event.pressed;
+  }
+
+  // Now let's transform these into the "cadet request" and "greek request."
+  const bool greek_request = (greek_held && !alt_held);
+  const bool cadet_request = (cadet_held || (greek_held && alt_held));
+
+  // Now, handle the lock keys. We store next_layer_lock in a local variable so that we can
+  // determine the layer to pick right now before we update layer_lock.
+  uint32_t next_layer_lock = layer_lock;
+  if (keycode == KC_CAPS) {
+    // If we're in QWERTY mode, caps lock is already going to be managed by the host OS, but by
+    // tracking it ourselves we can also usefully apply it to the GREEK and CADET layers.
+    if (record->event.pressed) {
+      shift_lock = !shift_lock;
+    }
+  } else if (keycode == KC_LAYER_LOCK) {
+    if (record->event.pressed) {
+      if (cadet_request) {
+        next_layer_lock = _CADET;
+      } else if (greek_request) {
+        next_layer_lock = _GREEK;
+      } else {
+        next_layer_lock = _QWERTY;
+      }
+    }
+  }
+
+  // OK! Now we know which buttons are being held, and the current and upcoming states of the locks.
+  // We can compute our new base layer. Remember that the CADET and GREEK keys act as their own
+  // antonyms if they match the layer lock -- e.g., if you have CADET locked, then CADET+X generates
+  // QWERTY-X.
+  uint32_t base_layer;
+  if (cadet_request) {
+    base_layer = (layer_lock == _CADET ? _QWERTY : _CADET);
+  } else if (greek_request) {
+    base_layer = (layer_lock == _GREEK ? _QWERTY : _GREEK);
+  } else {
+    base_layer = layer_lock;
+  }
+
+  const bool shifted = (shift_held != shift_lock);
+  uint32_t actual_layer;
+  if (base_layer == _CADET) {
+    actual_layer = (shifted ? _SHIFTCADET : _CADET);
+  } else if (base_layer == _GREEK) {
+    actual_layer = (shifted ? _SHIFTGREEK : _GREEK);
+  } else {
+    // We don't do shifting for the QWERTY layer, since for that we emit USB HID codes and shifting
+    // is managed by the host OS.
+    actual_layer = _QWERTY;
+  }
+
+  // And now we can update the layer lock and the actual firmware layer selector.
+  layer_lock = next_layer_lock;
+  uint32_t new_layer_state = (layer_state & LAYER_MASK) | (1UL << actual_layer);
```

Comment:
The layer mask isn't always 32 bit, and can be changed. 

If you could change to the typedef that we recently added? 


```suggestion
  layer_state_t new_layer_state = (layer_state & LAYER_MASK) | (1UL << actual_layer);
```

Reply:
Updated as per above.